lib/DhallToCabal/ConfigTree.hs

```diff
@@ -0,0 +1,172 @@
+{-# language DeriveFunctor #-}
+{-# language LambdaCase #-}
+{-# language OverloadedStrings #-}
+
+module DhallToCabal.ConfigTree ( ConfigTree(..), toConfigTree ) where
+
+import Control.Applicative
+import Control.Lens (deepOf)
+import Data.Functor.Identity
+import Data.Monoid
+import Dhall.Core hiding (Const)
+
+
+
+data ConfigTree cond a
+  = Leaf a
+  | Branch cond ( ConfigTree cond a ) ( ConfigTree cond a )
+  deriving (Functor, Show)
+
+
+
+toConfigTree :: ( Eq a, Eq s ) => Expr s a -> ConfigTree ( Expr s a ) ( Expr s a )
+toConfigTree e =
+  let
+    v =
+      "config"
+
+    saturated =
+      normalize (App e (Var v))
+
+    loop e =
+      let
+        Const condition =
+          deepOf
+            subExpr
+            ( findConfigUse v )
+            ( Const . First . Just ) e
+
+      in
+        case condition of
+          First Nothing ->
+            Leaf e
+
+          First (Just cond) ->
+            let
+              replaceWith x =
+                loop
+                  . normalize
+                  . runIdentity
```

Comment:
I think you can simplify this to:

```
replaceWith x = loop . normalize . set (deepOf subExpr (exactly cond) x
```

Reply:
I wrote these out "by hand" as I was hoping to avoid a `lens` dependency. However, `deepOf` looks moderately involved enough that I couldn't be bothered unpicking it, so I could do this. However, I think "find the first application and then rename all occurances" can actually be done in one traversal with `MonadState (Maybe (Expr ... ))`, so I probably can't use `set` and `preview`. I will look into this, more this file was mostly a brain dump